library/src/commonMain/kotlin/com/badoo/reaktive/observable/CombineLatest.kt

```diff
@@ -0,0 +1,231 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+
+fun <T, R> Collection<Observable<T>>.combineLatest(mapper: (List<T>) -> R): Observable<R> =
+    observableByEmitter { emitter ->
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+        val lock = newLock()
+        val dummy = Any()
+        val values = MutableList<Any?>(size) { dummy }
+        var readyValues: List<T>? = null
+        var activeSourceCount = size
+
+        forEachIndexed { index, source ->
+            source.subscribeSafe(
+                object : ObservableObserver<T> {
+                    override fun onSubscribe(disposable: Disposable) {
+                        disposables += disposable
+                    }
+
+                    override fun onNext(value: T) {
+                        lock.synchronized {
+                            values[index] = value
+                            if ((readyValues == null) && values.none { it === dummy }) {
+                                @Suppress("UNCHECKED_CAST")
+                                readyValues = values as List<T>
+                            }
+
+                            readyValues
+                                ?.let(mapper)
+                                ?.also(emitter::onNext)
+                        }
+                    }
+
+                    override fun onComplete() {
+                        lock.synchronized {
+                            if (--activeSourceCount == 0) {
+                                emitter.onComplete()
+                            }
+                        }
+                    }
+
+                    override fun onError(error: Throwable) {
+                        lock.synchronized {
```

Comment:
1. Lock is needed as there is a concurrency: multiple sources can emit errors simultaneously
2. As we are using emitter, it will dispose upstream on error. It is safe to emit after onError or onComplete, observableByEmiiter takes care of it.